payjoin/src/psbt/merge.rs

```diff
@@ -0,0 +1,268 @@
+use std::fmt;
+
+use bitcoin::{OutPoint, Psbt};
+
+#[allow(dead_code)]
+/// Error type for merging two unique unsigned PSBTs
+#[derive(Debug, PartialEq)]
+pub(crate) enum MergePsbtError {
+    /// Input from other PSBT already exists in this PSBT
+    InputAlreadyExists(OutPoint),
+    /// Input is already signed
+    MyInputIsSigned(OutPoint),
+    /// Other PSBT's input is already signed
+    OtherInputIsSigned(OutPoint),
+}
+
+impl fmt::Display for MergePsbtError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::InputAlreadyExists(outpoint) =>
+                write!(f, "input already exists with outpoint: {}", outpoint),
+            Self::MyInputIsSigned(outpoint) =>
+                write!(f, "my input is already signed with outpoint: {}", outpoint),
+            Self::OtherInputIsSigned(outpoint) =>
+                write!(f, "other input is already signed with outpoint: {}", outpoint),
+        }
+    }
+}
+
+impl std::error::Error for MergePsbtError {}
+
+#[allow(dead_code)]
+pub(crate) trait MergePsbtExt: Sized {
+    fn dangerous_clear_signatures(&mut self);
+    fn merge_unsigned_tx(&mut self, other: Self) -> Result<(), Vec<MergePsbtError>>;
+}
```

Comment:
The biggest design question I have with this commit is "Why a trait and not pure functions?"

The trait requires `&mut self` which is more restrictive than a `clone` and new Psbt being returned. What's the advantage I'm missing? Fewer imports (Trait vs individual methods or qualification?)

Reply:
> "Why a trait and not pure functions?"

It boils down one-to-one or many-to-many operations and which one is going to be a more commond usecase. Trait ext makes sense if we are working with two single PSBTs (which was the case during the 2s1r POC). However, with ns1r it would be more ergonomic if we just used a stateless function independent of any existing type. However `dangerous_clear_signatures` would make sense as an extention of the existing rust-bitcoin type

Reply:
I guess you're expressing a preference for the former over the latter? 

```rs
let merged = psbt1
  .merge(psbt2)
  .merge(psbt3)
```

instead of

```rs
let psbts: IntoIter<Psbt>;
let merged = psbts
  .iter()
  .reduce(merge);
```

?

Reply:
>I guess you're expressing a preference for the former over the latter?

I would actually prefer a pure function for merging `Iter<psbt>`.  This actually simplifies the merging operation as I only have to merge input/outputs in the unsigned txs and don't have to worry about psbt fields. I pushed a version of psbt merge as a stateless function in [ee3c882](https://github.com/payjoin/rust-payjoin/pull/523/commits/ee3c882dd5b14d6860b14cb2517feb1ef85db412)

Reply:
My one point of contention is what to do about incompatibly tx versions and nsequence locktimes. For now i've added some validation where we reject non-v2 txs and if the tx has a non-zero nsequence. 

Intrested to know what you think.

Reply:
>psbts
  .into_iter()
  .reduce(merge);

I like the reduce approach better than what I have now. 
```rust
fn merge_unsigned_tx(
    psbts: impl Iterator<Item = Psbt>,
) -> Result<Psbt, Vec<MergePsbtError>> {
... }
``` 

More idomatic.

Reply:
Yeah, I wasn't trying to imply that the basic util should take an iterator, but it should just do the merge operation and you could literally use reduce on an iterator in an actual implementation. I think it makes sense for a merge(psbt_a, psbt_b) operation to exist since it's testable on its own.

Reply:
The only problem is reduce is doesnt allow for errors to be returned. And maybe thats not such a bad thing. Merge doesnt really have to fail. It only fails b/c it enforces the business logic of ns1r. It would make sense to keep merge lean ` fn merge_unsigned_tx(acc: Psbt, psbt: Psbt) -> Psbt` and do this validation on the reciever.

Reply:
I didn't know that it can't return errors. Makes it worth using another design.